tls/s2n_async_pkey.c

```diff
@@ -393,19 +398,43 @@ S2N_RESULT s2n_async_pkey_sign_apply(struct s2n_async_pkey_op *op, struct s2n_co
 
     struct s2n_async_pkey_sign_data *sign = &op->op.sign;
 
+    /* Perform signature validation only if feature is buy in */
+    RESULT_ENSURE_REF(conn->config);
+    if (conn->config->async_pkey_validation_mode == S2N_ASYNC_PKEY_VALIDATION_STRICT) {
+        RESULT_ENSURE(!(s2n_async_pkey_verify_signature(conn, sign->sig_alg,
+                                                &sign->digest_copy, &sign->signature) < 0), S2N_ERR_VERIFY_SIGNATURE);
+    }
+
     RESULT_GUARD_POSIX(sign->on_complete(conn, &sign->signature));
 
     return S2N_RESULT_OK;
 }
 
+int s2n_async_pkey_verify_signature(struct s2n_connection *conn, s2n_signature_algorithm sig_alg,
+                                    struct s2n_hash_state *digest, struct s2n_blob *signature) {
+    POSIX_ENSURE_REF(conn);
+    POSIX_ENSURE_REF(conn->handshake_params.our_chain_and_key);
+
+    DEFER_CLEANUP(struct s2n_pkey public_key = {0}, s2n_pkey_free);
+    s2n_pkey_type pkey_type = S2N_PKEY_TYPE_UNKNOWN;
+    POSIX_GUARD(s2n_asn1der_to_public_key_and_type(&public_key, &pkey_type,
+            &conn->handshake_params.our_chain_and_key->cert_chain->head->raw));
```

Comment:
Allocating memory and parsing public key each time might end up being expensive, we should really look into doing it only once per certificate.